clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -1091,9 +1098,12 @@ Node selectReadReplica(TopicPartition partition, Node leaderReplica, long curren
     private Map<Node, FetchSessionHandler.FetchRequestData> prepareFetchRequests() {
```

Comment:
The `fetchablePartitions` method used below is probably another nice opportunity to use something like `forEachAssignedPartition`.

Reply:
I'm not sure it saves anything in this case. The main benefit of `forEachAssignedPartition` is avoiding making a copy of the assignment set. Since `fetchablePartitions` iterates across the internal set directly I don't think it would help

Reply:
We can avoid a copy in the case @hachikuji mentions as well, right? See below:

> synchronized List<TopicPartition> fetchablePartitions(Predicate<TopicPartition> isAvailable) {
>         return assignment.stream()
>                 .filter(tpState -> isAvailable.test(tpState.topicPartition()) && tpState.value().isFetchable())
>                 .map(PartitionStates.PartitionState::topicPartition)
>                 .collect(Collectors.toList());
>     }

Reply:
Ah ok, I see what he means. I'll look into this

Reply:
I made a pass at this and it wasn't so simple. Deferring for now